README.md
# express-web-file-store

## Usage

```js
import path from 'path'
import express from 'express'
import fileStore, { LocalStore, SeaweedFS } from 'express-web-file-store'

const store = fileStore({
  verbose: true
})

// 1. YourStore => failure => YourStore1 (default)
// 2. YourStore => parallel => YourStore1 (not implemented)
// store.use(new YourStore())
// store.use(new YourStore1())
store.use(new LocalStore({
  // Read-only addresses, can have multiple, sequential reads
  rootReadonly: ['/a/b/path'],
  // Read/Write address
  // Read Order: rootReadonly -> root
  // if root not set, Write will throw error
  root: path.join(__dirname, './upload'),
  // see https://www.expressjs.com.cn/4x/api.html#express.static
  static: {
    maxAge: 31536000 * 1000 // ms, default 0 ms
  }
}))

// use seaweedfs store https://github.com/chrislusf/seaweedfs
store.use(new SeaweedFS({
  filerServer: 'http://seaweedfs-filer:8888/',
  etag: true, // default true
  lastModified: true, // default true
  maxAge: 31536000 * 1000 // ms, default 0 ms
}))

const app = express()
// POST /filename multipart file
// DELETE /filename
// GET /filename
app.use('/', store)

app.listen(1234)

```

## License

[MIT](./LICENSE)


CHANGELOG.md
## [1.0.11](https://github.com/somewind/express-web-file-store/compare/v1.0.5...v1.0.11) (2020-03-18)


### Bug Fixes

* 404 compatibility ([733d4fb](https://github.com/somewind/express-web-file-store/commit/733d4fb))
* fix unlink copyFile error ([81ddb58](https://github.com/somewind/express-web-file-store/commit/81ddb58))
* tempfile upload no such file or directory error ([fba7f24](https://github.com/somewind/express-web-file-store/commit/fba7f24))


### Features

* SeaweedFS support cache ([79c7e1c](https://github.com/somewind/express-web-file-store/commit/79c7e1c))
* support SeaweedFS ([3703f2e](https://github.com/somewind/express-web-file-store/commit/3703f2e))



## [1.0.10](https://github.com/somewind/express-web-file-store/compare/v1.0.5...v1.0.10) (2020-03-18)


### Bug Fixes

* fix unlink copyFile error ([81ddb58](https://github.com/somewind/express-web-file-store/commit/81ddb58))
* tempfile upload no such file or directory error ([fba7f24](https://github.com/somewind/express-web-file-store/commit/fba7f24))


### Features

* SeaweedFS support cache ([79c7e1c](https://github.com/somewind/express-web-file-store/commit/79c7e1c))
* support SeaweedFS ([3703f2e](https://github.com/somewind/express-web-file-store/commit/3703f2e))



## [1.0.7](https://github.com/somewind/express-web-file-store/compare/v1.0.5...v1.0.7) (2020-01-08)


### Bug Fixes

* fix unlink copyFile error ([81ddb58](https://github.com/somewind/express-web-file-store/commit/81ddb58))
* tempfile upload no such file or directory error ([fba7f24](https://github.com/somewind/express-web-file-store/commit/fba7f24))



## [1.0.6](https://github.com/somewind/express-web-file-store/compare/v1.0.5...v1.0.6) (2020-01-07)


### Bug Fixes

* fix unlink copyFile error ([81ddb58](https://github.com/somewind/express-web-file-store/commit/81ddb58))



## [1.0.5](https://github.com/somewind/express-web-file-store/compare/v1.0.2...v1.0.5) (2019-12-26)


### Bug Fixes

* ERROR EXDEV: cross-device ([ac51096](https://github.com/somewind/express-web-file-store/commit/ac51096))
* ignore mkdir error ([3eec08f](https://github.com/somewind/express-web-file-store/commit/3eec08f))



## [1.0.3](https://github.com/somewind/express-web-file-store/compare/v1.0.2...v1.0.3) (2019-12-25)


### Bug Fixes

* ignore mkdir error ([3eec08f](https://github.com/somewind/express-web-file-store/commit/3eec08f))



## [1.0.2](https://github.com/somewind/express-web-file-store/compare/v1.0.1...v1.0.2) (2019-12-05)


### Bug Fixes

* io error ([e798148](https://github.com/somewind/express-web-file-store/commit/e798148))



## 1.0.1 (2019-11-06)


### Features

* first version ([4923bfa](https://github.com/somewind/express-web-file-store/commit/4923bfa))
* rootReadonly config ([0f7eed1](https://github.com/somewind/express-web-file-store/commit/0f7eed1))





src/SeaweedFS.js
import fs from 'fs'
import lodash from 'lodash'
import axios from 'axios'
import FormData from 'form-data'
import url from 'url'

const defaultOptions = {
  filerServer: 'http://seaweedfs-filer:8888/',
  etag: true, // default true
  lastModified: true, // default true
  maxAge: 0 // ms, default 0 ms
}

export default class SeaweedFS {
  constructor (options = {}) {
    this.options = lodash.merge({}, defaultOptions, options)
  }

  _getFileUrl (req) {
    const fileurl = new url.URL(req.params.filepath, this.options.filerServer)
    return fileurl.href
  }

  async get (req, res) {
    const fileUrl = this._getFileUrl(req)
    if (!fileUrl) {
      const e = new Error('fileUrl is null')
      e.status = 404
      e.msg = 'fileUrl is null'
      throw e
    }

    const response = await axios.get(fileUrl, {
      responseType: 'stream',
      validateStatus: null
    })
    if (response.status === 200) {
      res.set('Content-Type', response.headers['content-type'])
      res.set('Content-Length', response.headers['content-length'])
      // set cache
      if (this.options.etag) {
        res.set('Etag', response.headers['etag'])
      }
      if (this.options.lastModified) {
        res.set('Last-Modified', response.headers['last-modified'])
      }
      if (typeof this.options.maxAge === 'number' && this.options.maxAge !== 0) {
        res.set('Cache-Control', `public, max-age=${this.options.maxAge / 1000}`)
      }
      response.data.pipe(res)
    } else {
      const e = new Error(response.statusText)
      e.status = response.status
      e.msg = response.statusText
      throw e
    }
  }

  async put (req, res) {
    const fileUrl = this._getFileUrl(req)
    if (!fileUrl) {
      const e = new Error('fileUrl is null')
      e.status = 404
      e.msg = 'fileUrl is null'
      throw e
    }

    const [file] = Object.values(req.files)

    const formData = new FormData()
    formData.append('file', fs.createReadStream(file.tempFilePath))
    const response = await axios.post(fileUrl, formData, {
      headers: formData.getHeaders()
    })

    if (response.status !== 201) {
      const e = new Error(response.statusText)
      e.status = response.status
      e.msg = response.statusText
      throw e
    }
  }

  async delete (req, res) {
    const fileUrl = this._getFileUrl(req)
    if (!fileUrl) {
      const e = new Error('fileUrl is null')
      e.status = 404
      e.msg = 'fileUrl is null'
      throw e
    }

    const response = await axios.delete(fileUrl, {
      validateStatus: null
    })
    if (response.status !== 204) {
      const e = new Error(response.statusText)
      e.status = response.status
      e.msg = response.statusText
      throw e
    }
  }
}
